### PR TITLE
Query v2

### PR DESCRIPTION
**Pull request by AsafMah**

### CODE REVIEW

File: azkustodata/kql/query_parameters.go [hunk 1]
```diff
@@ -1,70 +1,74 @@
 package kql
 
 import (
-	"github.com/Azure/azure-kusto-go/azkustodata/types"
```
Comment (yogilad) [inline azkustodata/kql/query_parameters.go:4]: Why the rename?
The old name sounds more to point.

File: azkustodata/kql/query_parameters.go [hunk 2]
```diff
@@ -1,70 +1,74 @@
...
 		panic("Invalid parameter values. make sure to adhere to KQL entity name conventions and escaping rules.")
 	}
-	q.parameters[key] = value
+	q.parameters[key] = v
 	return q
 }
 
-func (q *Parameters) AddBool(key string, value bool) *Parameters {
-	return q.addBase(key, newValue(value, types.Bool))
+func (q *Parameters) AddBool(key string, v bool) *Parameters {
```
Comment (yogilad) [inline azkustodata/kql/query_parameters.go:30]: Same question - does this support nulls?

File: azkustodata/kusto.go [hunk 3]
```diff
@@ -12,12 +15,14 @@ import (
...
 )
 
+type Statement = *kql.Builder
+
 // queryer provides for getting a stream of Kusto frames. Exists to allow fake Kusto streams in tests.
 type queryer interface {
 	io.Closer
 	query(ctx context.Context, db string, query Statement, options *queryOptions) (execResp, error)
 	mgmt(ctx context.Context, db string, query Statement, options *queryOptions) (execResp, error)
-	queryToJson(ctx context.Context, db string, query Statement, options *queryOptions) (string, error)
```
Comment (yogilad) [inline azkustodata/kusto.go:20]: Why break this APIs Name?

File: azkustodata/kusto.go [hunk 4]
```diff
@@ -172,6 +177,86 @@ func (c *Client) Query(ctx context.Context, db string, query Statement, options
 	return iter, nil
 }
 
+func (c *Client) newV1(
```
Comment (yogilad) [inline azkustodata/kusto.go:180]: New is usually reserved for instance creation.
Why do you use it in query/command APIs?

The set of API names need working.
Possible [Mgmt, QueryV1, QueryV2, IterativeQuery] where Iterative is V2 implicitly.

File: azkustodata/kusto.go [hunk 5]
```diff
@@ -172,6 +177,86 @@ func (c *Client) Query(ctx context.Context, db string, query Statement, options
...
+
+	return queryv2.NewStreamingDataSet(ctx, res, capacity)
+}
+
+func (c *Client) newV2(ctx context.Context, db string, kqlQuery Statement, options []QueryOption) (*queryOptions, io.ReadCloser, error) {
+	ctx, cancel := contextSetup(ctx)
+
+	options = append(options, V2NewlinesBetweenFrames())
+	options = append(options, V2FragmentPrimaryTables())
+	options = append(options, ResultsErrorReportingPlacement(ResultsErrorReportingPlacementEndOfTable))
```
Comment (yogilad) [inline azkustodata/kusto.go:238]: Do you test that being last allows you to override these options no matter what the user did?
Any options conflicting with those you should forcibly disable or fail the query over?

File: azkustodata/kusto.go [hunk 6]
```diff
@@ -172,6 +177,86 @@ func (c *Client) Query(ctx context.Context, db string, query Statement, options
...
+		return nil, nil, err
+	}
+
+	res, err := conn.rawQuery(ctx, db, kqlQuery, opts)
+
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+	return opts, res, nil
```
Comment (yogilad) [inline azkustodata/kusto.go:257]: Where's the whole iterative dataset reader?

File: azkustodata/kusto.go [hunk 7]
```diff
@@ -172,6 +177,86 @@ func (c *Client) Query(ctx context.Context, db string, query Statement, options
...
+}
+
 func (c *Client) QueryToJson(ctx context.Context, db string, query Statement, options ...QueryOption) (string, error) {
```
Comment (yogilad) [inline azkustodata/kusto.go:260]: Should you not leave this API public?
People may have taken a dependency on it?

File: azkustodata/kusto.go [hunk 8]
```diff
@@ -200,12 +290,6 @@ func (c *Client) QueryToJson(ctx context.Context, db string, query Statement, op
 // Note that the server has a timeout of 10 minutes for a management call by default unless the context deadline is set.
 // There is a maximum of 1 hour.
 func (c *Client) Mgmt(ctx context.Context, db string, query Statement, options ...QueryOption) (*RowIterator, error) {
-	if stmt, ok := query.(Stmt); ok {
```
Comment (yogilad) [inline azkustodata/kusto.go:203]: Did this move somewhere?
Otherwise, why the change?

File: azkustodata/query/column.go [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package query
+
+import "github.com/Azure/azure-kusto-go/azkustodata/types"
+
+// Column represents a column in a table.
+type Column interface {
```
Comment (yogilad) [inline azkustodata/query/column.go:6]: Why would such a simple class need an interface?

File: azkustodata/query/from_kusto.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
+package query
+
+// value.go provides methods for converting a row to a *struct and for converting KustoValue into Go types
```
Comment (yogilad) [inline azkustodata/query/from_kusto.go:3]: update this (I did not look at other files, but they can also be out of date)

**Review by yogilad — COMMENTED** [review]
very partial review (read through some files)
Will carry on later

File: azkustodata/example_iterative_query_test.go [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+		if err != nil {
+			panic(err)
+		}
+	}(client)
+
+	ctx := context.Background()
+
+	// Simple query - single table
+
+	dataset, err := client.IterativeQuery(ctx, "Samples", kql.New("PopulationData"))
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/example_iterative_query_test.go:36]: This table was chosen because it only has 52 rows

File: azkustodata/example_mgmt_test.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+		if err != nil {
+			panic(err)
+		}
+	}(client)
+
+	ctx := context.Background()
+
+	// Simple query - single table
+
+	dataset, err := client.MgmtNew(ctx, "Samples", kql.New(".show tables"))
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/example_mgmt_test.go:30]: Mgmt is taken, it's there until I remove the old one.

File: azkustodata/example_mgmt_test.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+		println(row.Ordinal())
+
+		// Or iterate over all rows
+		for _, row := range rows {
+			// Each row has an index and a pointer to the table it belongs to
+			println(row.Ordinal())
+			println(row.Table().Name())
+
+			// There are a few ways to access the values of a row:
+			val := row.Value(0)
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/example_mgmt_test.go:67]: What's GetTypeByRef? The Value object has a "GetType" method if this is what you mean?

File: azkustodata/kql/builder.go [hunk 14]
```diff
@@ -43,6 +43,11 @@ func (b *Builder) addBase(value fmt.Stringer) *Builder {
 	return b
 }
 
+func (b *Builder) AddValue(v value.Kusto) *Builder {
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/kql/builder.go:46]: The name was here before me.  It means "Kusto Value" I guess.
We can call it value.Value or another name if you have it

File: azkustodata/kql/builder.go [hunk 15]
```diff
@@ -55,44 +60,48 @@ func (b *Builder) AddLiteral(value stringConstant) *Builder {
 	return b.addBase(value)
 }
 
-func (b *Builder) AddBool(value bool) *Builder {
-	return b.addBase(newValue(value, types.Bool))
+func (b *Builder) AddBool(v bool) *Builder {
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/kql/builder.go:63]: It doesn't.
The type bool doesn't support it, I can change it to accept a bool reference, but that isn't great either.

There is now an "AddValue" api you can add from a (nullable) kusto value:
b.AddValue(value.NewNullBool())

File: azkustodata/kql/query_parameters.go (hunk 1 above)
Comment (AsafMah) [reply azkustodata/kql/query_parameters.go:4]: If you have a "types.Bool" you don't have a bool type, you have a bool value. 
That was my logic, but I can revert it otherwise.

File: azkustodata/kql/query_parameters.go (hunk 2 above)
Comment (AsafMah) [reply azkustodata/kql/query_parameters.go:30]: Same answer

File: azkustodata/kusto.go (hunk 3 above)
Comment (AsafMah) [reply azkustodata/kusto.go:20]: 1. This is a private api
2. The method no longer returns a json string - it returns a raw response object.

File: azkustodata/kusto.go (hunk 6 above)
Comment (AsafMah) [reply azkustodata/kusto.go:257]: all of the "major" code is in query.v2

File: azkustodata/kusto.go (hunk 7 above)
Comment (AsafMah) [reply azkustodata/kusto.go:260]: This was an api that was specifically requested from us

File: azkustodata/query/column.go (hunk 9 above)
Comment (AsafMah) [reply azkustodata/query/column.go:6]: To not have breaking changes if the internals change.

File: azkustodata/example_mgmt_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (yogilad) [reply azkustodata/example_mgmt_test.go:67]: Sorry, Get[Type]By[Ref]() as in the set of APIs for _GetBooleanByOrdinal_, _GetStringByColumnName_ and so on

File: azkustodata/kql/builder.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (yogilad) [reply azkustodata/kql/builder.go:46]: What does this do?
Isn't this the interface for all KQL types?

File: azkustodata/example_mgmt_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (yogilad) [reply azkustodata/example_mgmt_test.go:30]: New implies a CTOR like behavior in Go.
Can you add a comment this is temporary and should not be committed?

File: azkustodata/kql/builder.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (yogilad) [reply azkustodata/kql/builder.go:63]: Why is that not great?
What happens when you get a type back? how do you indicate a row is null?

File: azkustodata/kql/query_parameters.go (hunk 1 above)
Comment (yogilad) [reply azkustodata/kql/query_parameters.go:4]: not sure I follow.
Those are data types. the do store values, but you code the type behavior.
Suggest to revert (it would also simplify this PR)

File: azkustodata/kusto.go (hunk 7 above)
Comment (yogilad) [reply azkustodata/kusto.go:260]: Not sure how this answers the question.
Can you clarify?

**Review by yogilad — COMMENTED** [review]
Just replies

File: azkustodata/kusto.go (hunk 7 above)
Comment (AsafMah) [reply azkustodata/kusto.go:260]: Not sure what you mean then - I didn't change this API

File: azkustodata/kusto.go (hunk 5 above)
Comment (AsafMah) [reply azkustodata/kusto.go:238]: It will override all of the options.
I can make it fail if they define progressive=true if you want

File: azkustodata/kusto.go (hunk 4 above)
Comment (AsafMah) [reply azkustodata/kusto.go:180]: Again, the conflict with the old names

File: azkustodata/kql/builder.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/kql/builder.go:63]: Because they now can't do AddBool(true), they have to define a variable and make it into a reference.

When returning it's simpler

File: azkustodata/kql/query_parameters.go (hunk 1 above)
Comment (AsafMah) [reply azkustodata/kql/query_parameters.go:4]: My mistake - I didn't change any namespace, they were named like this before. 
the types package defines the column types (like types.Bool is a constant that means "bool"), and the value defines a value (like value.Bool is a struct that holds a kusto bool ).

The difference here is now I need a different improt

File: azkustodata/kusto.go (hunk 8 above)
Comment (AsafMah) [reply azkustodata/kusto.go:203]: not sure why, I'll revert - but this whole method is going to be removed soon anyway.

File: azkustodata/query/from_kusto.go (hunk 10 above)
Comment (AsafMah) [reply azkustodata/query/from_kusto.go:3]: It's only for this file, which was taken from the old system.

File: azkustodata/example_iterative_query_test.go [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	// The tables are streamed, so you need to iterate through the `Results()` channel to get them
+
+	for tableResult := range dataset.Results() {
+		// Make sure to always check for errors
+		if tableResult.Err() != nil {
+			// It's possible to get an errors and still get a table - partial results
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/example_iterative_query_test.go:47]: In fatal errors there will be no table results.

File: azkustodata/example_iterative_query_test.go [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+		// 2. GetAllRows() - reads all rows and closes the table
+		rows, errors := table.GetAllRows()
+		for _, row := range rows {
+			println(row.Ordinal())
+		}
+		for _, err := range errors {
+			println(err.Error())
+		}
+
+		// 3. Rows() - reads rows as they are received
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/example_iterative_query_test.go]: Fixed

File: azkustodata/example_iterative_query_test.go [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+			err = row.ToStruct(&pd)
+			if err != nil {
+				panic(err)
+			}
+			println(pd.State)
+			println(pd.Pop)
+
+		}
+
+		// Or you can easily get the results as a slice of structs
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/example_iterative_query_test.go]: Fixed

File: azkustodata/example_mgmt_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/example_mgmt_test.go:30]: Added

File: azkustodata/example_mgmt_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/example_mgmt_test.go:79]: Oh yes, I'll add the example

File: azkustodata/kql/builder.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply azkustodata/kql/builder.go:46]: Yes, what would you call it?

File: azkustodata/query/v2/iterative_dataset.go [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+		rows, err := table.GetAllRows()
+		tables = append(tables, query.NewDataTable(d, table.Ordinal(), table.Id(), table.Name(), table.Kind(), table.Columns(), rows, err))
+	}
+
+	return tables, errs
+}
+
+// NewStreamingDataSet creates a new iterativeDataset from a reader.
+// The capacity parameter is the capacity of the channel that receives the frames from the Kusto service.
+func NewStreamingDataSet(ctx context.Context, r io.ReadCloser, capacity int) (IterativeDataset, error) {
```
Comment (yogilad) [inline azkustodata/query/v2/iterative_dataset.go:104]: Itraerative

File: azkustodata/query/v2/json.go [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+
+// readFramesIterative reads frames from a reader and sends them to a channel as they are read.
+func readFramesIterative(reader io.Reader, ch chan<- Frame) error {
+	defer close(ch)
+
+	// Crazily enough, json.Decoder always puts THE ENTIRE READER IN MEMORY
+	// So we have to manually split the reader into lines and decode each line with a new decoder
+
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
```
Comment (yogilad) [inline azkustodata/query/v2/json.go:140]: See if this duplicates the data from the input stream?
What does slice do?

File: azkustodata/value/value.go [hunk 21]
```diff
@@ -35,15 +35,47 @@ The Unmarshal() is for internal use, it should not be needed by an end user. Use
...
+	case types.Int:
+		return NewNullInt()
+	case types.Long:
+		return NewNullLong()
+	case types.Real:
+		return NewNullReal()
+	case types.Decimal:
+		return NewNullDecimal()
+	case types.String:
+		return NewNullString()
```
Comment (yogilad) [inline azkustodata/value/value.go:67]: No null in string

File: azkustodata/query/table.go [hunk 22]
```diff
@@ -0,0 +1,29 @@
+package query
+
+import "github.com/Azure/azure-kusto-go/azkustodata/errors"
+
+type Table interface {
+	Id() string
+	Index() int64
+	Name() string
+	Columns() []Column
+	Kind() string
```
Comment (yogilad) [inline azkustodata/query/table.go:10]: enum or consts

File: azkustodata/query/table_impl.go [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func NewFullTable(base Table, rows []Row) FullTable {
+	return &fullTable{
+		Table: base,
+		rows:  rows,
+	}
+}
+
+func ReplaceFullTableRows(t FullTable, rows []Row) FullTable {
```
Comment (yogilad) [inline azkustodata/query/table_impl.go:85]: reminder: row ref to col instead of table and maybe this will become redundant

File: azkustodata/query/v2/full_dataset.go [hunk 24]
```diff
@@ -0,0 +1,40 @@
+package v2
+
+import (
+	"context"
+	"github.com/Azure/azure-kusto-go/azkustodata/errors"
+	"github.com/Azure/azure-kusto-go/azkustodata/query"
+)
+
+// FullDataset represents a full result from kusto - where all the tables are received before the dataset is returned.
+type fullDataset struct {
```
Comment (yogilad) [inline azkustodata/query/v2/full_dataset.go:10]: Check if this is a duplicate of the shared impl
